### PR TITLE
Adds a super basic multiz power bridge

### PR DESCRIPTION
**Pull request by Kmc2000**
## About The Pull Request

This PR is a quick port of something I did on ST-DS13.

## Why It's Good For The Game

On my server I needed a way to power multiple decks, this does just that. It'll bridge the powernets of your multiZ layers.

Kev asked me to port any potentially useful multiZ stuff so I'll submit this for consideration.

All you need to do is place a cable node (knot) under it and connect it to a powernet, then ensure it lines up with ones on desired decks. If it breaks in game, use a multitool on it to regenerate its connections.

If you wish, I can make the sprite a bit more industrial but this is what it is currently:

On:

Off:

## Changelog
:cl: Kmc2000
add: Added a basic multiZ power relay
/:cl:

### CODE REVIEW

**Comment by ShizCalev** [comment]
so this indiscriminately powers a relay across all z's as long as they're lined up, yeah?

**Comment by Kmc2000** [comment]
Yeah, long as it's lined up it'll work.

**Comment by ShizCalev** [comment]
Should probably check to make sure the Z's it's accessible to share the same Zlevel trait, that way we don't end up with someone trying to power a random space ruin or something over at lavaland from the station (in the event they ever become construct-able)

File: code/modules/power/multiz_cable.dm [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/obj/machinery/power/deck_relay //This bridges powernets
+	name = "Multi-deck power adapter"
+	desc = "This impressive machine uses plasma based power transmission technology to supply power to alternate decks. It requires a steady power source. Click it with a multitool / ODN scanner to see stats and reacquire connections."
+	icon = 'icons/obj/power.dmi'
+	icon_state = "cablerelay-off"
+	var/obj/machinery/power/deck_relay/below
+	var/obj/machinery/power/deck_relay/above
+	var/list/relays = list() //to bridge the powernets.
+	anchored = TRUE
+	density = FALSE
```
Comment (ShizCalev) [inline code/modules/power/multiz_cable.dm]: move these above the new vars

File: code/modules/power/multiz_cable.dm [hunk 2]
```diff
@@ -0,0 +1,65 @@
+/obj/machinery/power/deck_relay //This bridges powernets
+	name = "Multi-deck power adapter"
+	desc = "This impressive machine uses plasma based power transmission technology to supply power to alternate decks. It requires a steady power source. Click it with a multitool / ODN scanner to see stats and reacquire connections."
+	icon = 'icons/obj/power.dmi'
+	icon_state = "cablerelay-off"
+	var/obj/machinery/power/deck_relay/below
+	var/obj/machinery/power/deck_relay/above
+	var/list/relays = list() //to bridge the powernets.
+	anchored = TRUE
```
Comment (ShizCalev) [inline code/modules/power/multiz_cable.dm]: .

File: code/modules/power/multiz_cable.dm [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	density = FALSE
+
+/obj/machinery/power/deck_relay/process()
+	refresh() //Sometimes the powernets get lost, so we need to keep checking.
+	if(powernet && (powernet.avail <= 0))		// is it powered?
+		icon_state = "cablerelay-off"
+	else
+		icon_state = "cablerelay-on"
+	if(!below || QDELETED(below) || !above || QDELETED(above))
```
Comment (ShizCalev) [inline code/modules/power/multiz_cable.dm]: !below & !above not existing would already be handled by `QDELETED`

File: code/modules/power/multiz_cable.dm [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		icon_state = "cablerelay-off"
+		find_relays()
```
Comment (ShizCalev) [inline code/modules/power/multiz_cable.dm]: this seems rather excessive to run every tick of process if either above or below aren't present.

File: code/modules/power/multiz_cable.dm [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+/obj/machinery/power/deck_relay/proc/find_relays()
+	relays = list()
+	var/turf/T = get_turf(src)
+	if(!T || !istype(T))
+		return FALSE
+	below = null //in case we're re-establishing
+	var/obj/structure/cable/C = T.get_cable_node() //check if we have a node cable on the machine turf, the first found is picked
+	if(C && C.powernet)
+		C.powernet.add_machine(src) //Nice we're in.
+		powernet = C.powernet
```
Comment (ShizCalev) [inline code/modules/power/multiz_cable.dm]: isn't this all done by `connect_to_network()`?

**Comment by nicbn** [comment]
It's odd to have to use a power bridge just to take cables upstairs. We should add upwards and downwards cables (and pipes as well).

**Comment by Kmc2000** [comment]
When you code it 😎

File: code/modules/power/multiz_cable.dm [hunk 6]
```diff
@@ -3,19 +3,19 @@
...
-	density = FALSE
 
 /obj/machinery/power/deck_relay/process()
 	refresh() //Sometimes the powernets get lost, so we need to keep checking.
 	if(powernet && (powernet.avail <= 0))		// is it powered?
 		icon_state = "cablerelay-off"
 	else
 		icon_state = "cablerelay-on"
-	if(!below || QDELETED(below) || !above || QDELETED(above))
+	if(!below || !above)
```
Comment (ShizCalev) [inline code/modules/power/multiz_cable.dm]:
```suggestion
		if(QDELETED(below) || QDELETED(above))
```

**Comment by Nickkoranda** [comment]
Are these craftable or only for mappers?

**Comment by Kmc2000** [comment]
Mapping only, it's just a basic tool if anyone wants to experiment with multiZ

File: code/modules/power/multiz_cable.dm (hunk 4 above)
Comment (optimumtact) [reply code/modules/power/multiz_cable.dm]: this should not be a tick at all, just have a global listener for deck relay creation and notify all nearby relays at the time of creation/deletion

File: code/modules/power/multiz_cable.dm [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+/obj/machinery/power/deck_relay/proc/refresh()
+	var/turf/ours = get_turf(src)
+	for(var/X in relays) //Typeless loops are apparently more efficient.
+		var/obj/machinery/power/deck_relay/DR = X
+		if(!DR)
+			continue
+		var/turf/T = get_turf(DR)
+		var/obj/structure/cable/C = T.get_cable_node()
+		var/obj/structure/cable/XR = ours.get_cable_node()
+		if(C && XR)
```
Comment (SpaceManiac) [inline code/modules/power/multiz_cable.dm]: Minor, but could save a proccall here with ` && C != XR`

**Review by SpaceManiac — CHANGES_REQUESTED** [review]

File: code/modules/power/multiz_cable.dm [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+			merge_powernets(XR.powernet,C.powernet)//Bridge the powernets.
+
+/obj/machinery/power/deck_relay/proc/find_relays()
+	relays = list()
+	var/turf/T = get_turf(src)
+	if(!T || !istype(T))
+		return FALSE
+	below = null //in case we're re-establishing
+	above = null
```
Comment (SpaceManiac) [inline code/modules/power/multiz_cable.dm]: Both these `= null` assignments are never used and are immediately overwritten by the following two lines.

**Pull request closed without merging**